tla-io/src/main/scala/at/forsyte/apalache/io/tnt/Tnt.scala

```diff
@@ -0,0 +1,149 @@
+package at.forsyte.apalache.io.tnt
+
+import scala.util.Try
+import scala.collection.mutable
+
+import com.typesafe.scalalogging.LazyLogging
+
+import at.forsyte.apalache.tla.lir.{
+  BoolT1, ConstT1, FunT1, IntT1, OperT1, RecRowT1, RowT1, SeqT1, SetT1, StrT1, TlaEx, TlaType1, TupT1, VarT1, VariantT1,
+}
+
+object Tnt {
+
+  private[tnt] def exToTla(d: TntEx): Try[TlaEx] = null
+
+  // Convert a TntType into a TlaType1
+  //
+  // We implement a small family of mutually recursive conversion functions using this
+  // class in order to:
+  //
+  // - Encapsulate and store benign state used to track variable numbers (see below)
+  // - Support and encapsulate the mutual recursion needed in the methods
+  private[tnt] class typeToTlaType() extends LazyLogging {
+
+    // TNT type variables are represented with strings, but in TlaType1 they are integers.
+    // This little bundle of state lets us track the conversion of TNT var names to
+    // TlaType1 var numbers.
+    //
+    // Since the scope of type variables in TNT is always limited to single type-level type expressions,
+    // and since the converter class is constructed fresh for each (top-level) type conversion,
+    // we don't need to worry about variable name collisions.
+    var varNo = 0
+    val vars = mutable.Map[String, Int]()
+    def getVarNo(varName: String): Int = {
+      vars.get(varName) match {
+        case None    => val v = varNo; varNo += 1; v
+        case Some(n) => n
+      }
+    }
+
+    import TntType._
+
+    def rowToTupleElems(row: Row): List[TlaType1] = {
+      // Since we have to concat lists here, which can be expensive, we use an
+      // accumulator to benefit from tail call optimization. This is purely a precaution.
+      def aux(r: Row, acc0: List[TlaType1]): List[TlaType1] = r match {
+        // TODO Update with support for row-based tuples: https://github.com/informalsystems/apalache/issues/1591
+        // Row variables in tuples are not currently supported. But we will proceed assuming that this is an
+        // over generalization, and that we can safely treat the tuple as a closed tuple type.
+        case Row.Var(_) =>
+          logger.debug(
+              s"Found open row variable in TNT tuple with fields $row. Polymorphic tuples are not supported, but we will proceed assuming the tuple can be treated as a closed type.")
+          acc0
```

Comment:
@konnov and @gabrielamafra I thought of this approach after failing to figure out how sparse tuples could help :sweat_smile: 

I'm not sure if it will actually go thru :sweat_smile: I figure it's worth a shot, unless someone knows prima facie why this won't work. If so, I can convert this into an unsupported form, as we discussed, or pursue a different approach if someone can suggest one.

Reply:
Yep, it definitely makes sense to me that there are some TNT specs we won't be able to typecheck with this approach unless the operators are annotated.